devito/expression.py

```diff
@@ -46,4 +57,29 @@ def signature(self):
 
         :returns: List of unique data objects required by the expression
         """
-        return self.functions
+        return filter_ordered([f for f in self.functions],
+                              key=attrgetter('name'))
+
+    def indexify(self):
+        """Convert stencil expression to "indexed" format"""
+        self.stencil = dse_indexify(self.stencil)
+
+    @property
+    def index_offsets(self):
+        """Collect all non-zero offsets used with each index in a map
+
+        Note: This assumes we have indexified the stencil expression."""
+        offsets = defaultdict(list)
+        for e in preorder_traversal(self.stencil):
```

Comment:
I think you'll soon be able to use this: https://github.com/opesci/devito/blob/from-cse-to-dse/devito/symbolics.py#L339 to get Indexed thus avoiding a nest level (it needs generalisation though).
Also: for ``Indexed``, ``e.args[1:]`` is actually ``e.index`` or ``e.indices`` (I'm sure there's an attribute)